File: Baekjoon/2504.cpp
#include<iostream>
#include<stack>
#include<string>
using namespace std;

int main(){
    ios::sync_with_stdio(false);
    string str;
    stack<char> stk;
    int answer=0;
    int tmp=1;
    cin>>str;

    for(int i=0; i<str.length(); i++){
        if(str[i]=='('){
            tmp*=2;
            stk.push(str[i]);
        }
        else if(str[i]=='['){
            tmp*=3;
            stk.push(str[i]);
        }
        else if(str[i]==')'){
            if(stk.empty()) stk.push(str[i]);
            else if(str[i-1]=='('){ //묶음 계산
                answer+=tmp;
                stk.pop();
                tmp/=2;
            }
            else if(stk.top()=='('){ //이미 계산 완료
                stk.pop();
                tmp/=2;
            }
        }
        else if(str[i]==']'){
            if(stk.empty()) stk.push(str[i]);
            else if(str[i-1]=='['){
                answer+=tmp;
                stk.pop();
                tmp/=3;
            }
            else if(stk.top()=='['){
                stk.pop();
                tmp/=3;
            }
        }
    }

    if(stk.empty()) cout<<answer<<'\n';
    else cout<<0<<'\n';
    
    return 0;
}

File: Baekjoon/2504.py
#괄호의 값
s = input()
stk = []
tmp = 1
ans = 0

for i in range(len(s)):
    if(s[i] == '('):
        tmp *= 2
        stk.append(s[i])
    elif(s[i] == '['):
        tmp*=3
        stk.append(s[i])
    elif(s[i] == ')'):
        if not stk : stk.append(s[i])
        elif s[i-1] == '(':
            ans+=tmp
            stk.pop()
            tmp//=2 #몫
        elif stk[-1] == '(':
            stk.pop()
            tmp//=2
    elif(s[i] == ']'):
        if not stk : stk.append(s[i])
        elif s[i-1] == '[':
            ans+=tmp
            stk.pop()
            tmp//=3
        elif stk[-1] == '[':
            stk.pop()
            tmp//=3

if not stk:
    print(ans)
else:
    print(0)

